lightning/src/chain/package.rs

```diff
@@ -0,0 +1,1037 @@
+	preimage: PaymentPreimage,
+	htlc: HTLCOutputInCommitment
+}
+
+impl CounterpartyOfferedHTLCOutput {
+	pub(crate) fn build(per_commitment_point: PublicKey, counterparty_delayed_payment_base_key: PublicKey, counterparty_htlc_base_key: PublicKey, preimage: PaymentPreimage, htlc: HTLCOutputInCommitment) -> Self {
+		CounterpartyOfferedHTLCOutput {
+			per_commitment_point,
+			counterparty_delayed_payment_base_key,
+			counterparty_htlc_base_key,
+			preimage,
+			htlc
+		}
+	}
+}
+
+impl_writeable!(CounterpartyOfferedHTLCOutput, 33*3 + 32 + HTLC_OUTPUT_IN_COMMITMENT_SIZE, {
+	per_commitment_point,
+	counterparty_delayed_payment_base_key,
+	counterparty_htlc_base_key,
+	preimage,
+	htlc
+});
+
+/// A struct to describe a HTLC output on a counterparty commitment transaction.
+///
+/// HTLCOutputInCommitment (hash, timelock, directon) and pubkeys are used to generate a suitable
+/// witnessScript.
+#[derive(Clone, PartialEq)]
+pub(crate) struct CounterpartyReceivedHTLCOutput {
+	per_commitment_point: PublicKey,
+	counterparty_delayed_payment_base_key: PublicKey,
+	counterparty_htlc_base_key: PublicKey,
+	htlc: HTLCOutputInCommitment
+}
+
+impl CounterpartyReceivedHTLCOutput {
+	pub(crate) fn build(per_commitment_point: PublicKey, counterparty_delayed_payment_base_key: PublicKey, counterparty_htlc_base_key: PublicKey, htlc: HTLCOutputInCommitment) -> Self {
+		CounterpartyReceivedHTLCOutput {
+			per_commitment_point,
+			counterparty_delayed_payment_base_key,
+			counterparty_htlc_base_key,
+			htlc
+		}
+	}
+}
+
+impl_writeable!(CounterpartyReceivedHTLCOutput, 33*3 + HTLC_OUTPUT_IN_COMMITMENT_SIZE, {
+	per_commitment_point,
+	counterparty_delayed_payment_base_key,
+	counterparty_htlc_base_key,
+	htlc
+});
+
+/// A struct to describe a HTLC output on holder commitment transaction.
+///
+/// Either offered or received, the amount is always used as part of the bip143 sighash.
+/// Preimage is only included as part of the witness in former case.
+#[derive(Clone, PartialEq)]
+pub(crate) struct HolderHTLCOutput {
+	preimage: Option<PaymentPreimage>,
+	amount: u64,
+}
+
+impl HolderHTLCOutput {
+	pub(crate) fn build(preimage: Option<PaymentPreimage>, amount: u64) -> Self {
+		HolderHTLCOutput {
+			preimage,
+			amount
+		}
+	}
+}
+
+impl_writeable!(HolderHTLCOutput, 0, {
+	preimage,
+	amount
+});
+
+/// A struct to describe the channel output on the funding transaction.
+///
+/// witnessScript is used as part of the witness redeeming the funding utxo.
+#[derive(Clone, PartialEq)]
+pub(crate) struct HolderFundingOutput {
+	funding_redeemscript: Script,
+}
+
+impl HolderFundingOutput {
+	pub(crate) fn build(funding_redeemscript: Script) -> Self {
+		HolderFundingOutput {
+			funding_redeemscript,
+		}
+	}
+}
+
+impl_writeable!(HolderFundingOutput, 0, {
+	funding_redeemscript
+});
+
+/// A wrapper encapsulating all in-protocol differing outputs types.
+///
+/// The generic API offers access to outputs attributes (amount, weight, output_type) or allow
```

Comment:
This seems like a circular definition. If each variant is an output type, then it can't have an output type as an attribute. I'd drop this list entirely and say "common attributes"

I think instead you want to rename `output_type_eq` to something descriptive like `is_compatible` (or whatever is most accurate) -- best way to figure out how to name something is to first document it.